File: codewars/6kyu/backspaces-in-string/dduck-backspaces-in-string.py
# https://www.codewars.com/kata/backspaces-in-string/


def clean_string(s='', position=0):
    if position >= len(s):
        return s

    if s[position] != '#':
        return clean_string(s, position + 1)

    return clean_string(s[:max(0, position - 1)] + s[position + 1:], max(0, position - 1))


print(clean_string('abc#d##c'), "ac")
print(clean_string('abc####d##c#'), "?")
print(clean_string("#######"), "?")
print(clean_string(""), "?")
print(-1 or 0)

File: codewars/6kyu/backspaces-in-string/dduck-backspaces-in-string.js
// https://www.codewars.com/kata/backspaces-in-string/

function cleanString(s = '', position = 0) {
  if (position >= s.length) {
    return Array.isArray(s) ? s.join('') : s;
  }
  return s[position] !== '#' ? cleanString(s, position + 1)
    : cleanString(
      [
        ...s.slice(0, Math.max(0, position - 1)),
        ...s.slice(position + 1),
      ],
      Math.max(0, position - 1),
    );
}

console.log(cleanString('abc#d##c'), "ac");
console.log(cleanString('abc####d##c#'), "");
console.log(cleanString(''), "");
